command/agent/config_parse.go

```diff
@@ -168,3 +238,29 @@ func extraKeys(c *Config) error {
 
 	return helper.UnusedKeys(c)
 }
+
+// hcl.Decode will error if the ClientTemplateConfig isn't initialized with empty
+// structs, however downstream code expect nils if the struct only contains fields
+// with the zero value for its type. This function nils out type members that are
+// structs where all the member fields are just the zero value for its type.
+func finalizeClientTemplateConfig(config *Config) {
+	if config.Client.TemplateConfig.Wait.IsEmpty() {
+		config.Client.TemplateConfig.Wait = nil
+	}
+
+	if config.Client.TemplateConfig.WaitBounds.IsEmpty() {
+		config.Client.TemplateConfig.WaitBounds = nil
+	}
+
+	if config.Client.TemplateConfig.ConsulRetry.IsEmpty() {
+		config.Client.TemplateConfig.ConsulRetry = nil
+	}
+
+	if config.Client.TemplateConfig.VaultRetry.IsEmpty() {
+		config.Client.TemplateConfig.VaultRetry = nil
+	}
+
+	if config.Client.TemplateConfig.IsEmpty() {
+		config.Client.TemplateConfig = nil
+	}
+}
```

Comment:
yeah that makes sense.

Reply:
I just gave this a pass and ended up reverting. Moving this logic into the template runner resulted in having to make a call to a method that made a copy of this struct and then execute the `finalize` logic for every runner instance to avoid mutating shared state. This seemed wasteful. 

That said, I'd enjoy a discussion if you want/have time to have one. On the one hand, I don't love having domain logic related to a subsystem in the config loading/parsing code as that feels like a separate concern. On the other hand, I can understand an argument that that is actually the proper place as it localizes validation/initialization, and also because we only set these values to begin with because `hcl.decode` needs them to function properly. In the best case, we wouldn't even set them, and thus we wouldn't need to unset them.